### PR TITLE
LWDEV-9497 [OAuth] Check client authn type

### PR DESCRIPTION
**Pull request by MikhailGerasimov**

### CODE REVIEW

File: src/Lykke.Service.OAuth/Providers/AuthorizationProvider.cs [hunk 1]
```diff
@@ -106,6 +107,19 @@ public override async Task ValidateAuthorizationRequest(ValidateAuthorizationReq
...
             }
 
+
+            if (application.AuthorizationGrant != AuthorizationGrant.CodeOrImplicit)
+            {
+                if (context.Request.IsAuthorizationCodeFlow() && application.AuthorizationGrant != AuthorizationGrant.Code)
+                {
+                    context.Reject(OpenIdConnectConstants.Errors.InvalidRequest, $"AuthorizationCode flow not supported for client {context.ClientId}");
+                }
+                else if (context.Request.IsAuthorizationCodeFlow() && application.AuthorizationGrant != AuthorizationGrant.Implicit)
```
Comment (simpllfy) [inline src/Lykke.Service.OAuth/Providers/AuthorizationProvider.cs]: context.Request.IsAuthorizationCodeFlow() -> context.Request.IsImpicitFlow()

File: src/Lykke.Service.OAuth/Providers/AuthorizationProvider.cs [hunk 2]
```diff
@@ -106,6 +107,19 @@ public override async Task ValidateAuthorizationRequest(ValidateAuthorizationReq
                 return;
             }
 
+
+            if (application.AuthorizationGrant != AuthorizationGrant.CodeOrImplicit)
```
Comment (simpllfy) [inline src/Lykke.Service.OAuth/Providers/AuthorizationProvider.cs]: Should we check that request actually is implicit or auth code? or it's guaranteed by a server?

Comment (MikhailGerasimov) [reply src/Lykke.Service.OAuth/Providers/AuthorizationProvider.cs]: We check this here https://github.com/LykkeCity/Lykke.Service.OAuth/pull/71/files/2dec1db014c8663be43c3e54419bbe6ef539fab7#diff-dbcb6fc4447692ac295b968f09ebec03R75

File: src/Lykke.Service.OAuth/Providers/AuthorizationProvider.cs [hunk 3]
```diff
@@ -163,6 +180,22 @@ public override async Task ValidateTokenRequest(ValidateTokenRequestContext cont
                 return;
             }
 
+            var application = await _applicationRepository.GetByIdAsync(context.ClientId);
+            if (application.GrantTypes.Any())
+            {
+                if (context.Request.IsAuthorizationCodeGrantType() && !application.GrantTypes.Contains(GrantType.AuthorizationCode))
+                {
+                    context.Reject(OpenIdConnectConstants.Errors.UnsupportedGrantType, "AuthorizationCode grand type not supported for the client");
```
Comment (simpllfy) [inline src/Lykke.Service.OAuth/Providers/AuthorizationProvider.cs]: GranD -> GranT

File: src/Lykke.Service.OAuth/Providers/AuthorizationProvider.cs [hunk 4]
```diff
@@ -163,6 +180,22 @@ public override async Task ValidateTokenRequest(ValidateTokenRequestContext cont
...
+                    return;
+                }
+                if (context.Request.IsRefreshTokenGrantType() && !application.GrantTypes.Contains(GrantType.RefreshToken))
+                {
+                    context.Reject(OpenIdConnectConstants.Errors.UnsupportedGrantType, "Refresh token grand type not supported for client the client");
```
Comment (simpllfy) [inline src/Lykke.Service.OAuth/Providers/AuthorizationProvider.cs]: GranD -> GranT

File: src/Core/GrantType.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+﻿using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+
+namespace Core
+{
+    [JsonConverter(typeof(StringEnumConverter))]
+    public enum GrantType
```
Comment (gmaf) [inline src/Core/GrantType.cs]: I still dislike this approach. Why we can't use type-safe-enum pattern?
http://www.javacamp.org/designPattern/enum.html
https://stackoverflow.com/questions/424366/string-representation-of-an-enum

We could simply use ASOS to fill values in type-safe-enum.
https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/blob/85fb8c49a42dcbfe16042f8c17e9fc38177712f1/src/AspNet.Security.OpenIdConnect.Primitives/OpenIdConnectConstants.cs#L128-L134

I suggest is to store values for GrantType as array of strings like here:
image
https://developer.okta.com/docs/api/resources/oauth-clients#client-application-properties

Pros:
* Grant types in our database would be consistent with grant type strings provided by Oauth specification.
* Eliminate need of mapping our enum values to OAuth specification values
* Less code to check if client has context GrantType:
```cs
application.GrantTypes.Any(s => s.Equals(context.Request.GrantType, StringComparison.Ordinal));
```

**Pull request merged**